lib/active_admin/dynamic_setting.rb

```diff
@@ -0,0 +1,30 @@
+module ActiveAdmin
+
+  class DynamicSetting
+    def self.build(setting, type)
+      (type ? klass(type) : self).new(setting)
+    end
+
+    def self.klass(type)
+      klass = "#{type.to_s.camelcase}Setting"
+      raise ArgumentError, "Unknown type: #{type}" unless ActiveAdmin.const_defined?(klass)
+      ActiveAdmin.const_get(klass)
+    end
+
+    def initialize(setting)
+      @setting = setting
+    end
+
+    def value(*_args)
+      @setting
+    end
+  end
+
+  class StringSymbolOrProcSetting < DynamicSetting
+    def value(*args)
+      return @setting if @setting.is_a?(String) || @setting.nil?
+      (context = args.first) ? context.instance_eval(&@setting) : @setting.call
```

Comment:
@varyonic  would be nice to have yard docs or something like this for this method definition